src/mysqlsh_helpers.py

```diff
@@ -123,15 +208,56 @@ def configure_instance(self) -> None:
             self._wait_until_mysql_connection()
         except subprocess.CalledProcessError as e:
             logger.exception(f"Failed to configure instance: {self.instance_address}", exc_info=e)
-            raise MySQLInstanceConfigureError(e.stderr)
+            raise MySQLConfigureInstanceError(e.stderr)
 
     def create_cluster(self) -> None:
-        """Create an InnoDB cluster with Group Replication enabled."""
-        pass
+        """Create an InnoDB cluster with Group Replication enabled.
+
+        Raises MySQLCreateClusterError if there was an issue creating the cluster.
+        """
+        commands = (
+            f"shell.connect('{self.server_config_user}:{self.server_config_password}@{self.instance_address}')",
+            f"dba.create_cluster('{self.cluster_name}')",
+        )
+
+        try:
+            logger.debug("Creating a MySQL InnoDB cluster")
+            self._run_mysqlsh_script("\n".join(commands))
+        except subprocess.CalledProcessError as e:
+            logger.exception(
+                f"Failed to create cluster on instance: {self.instance_address}", exc_info=e
+            )
+            raise MySQLCreateClusterError(e.stderr)
 
-    def add_instance_to_cluster(self) -> None:
-        """Add an instance to the InnoDB cluster."""
-        pass
+    def add_instance_to_cluster(self, instance_address) -> None:
+        """Add an instance to the InnoDB cluster.
+
+        Raises MySQLADDInstanceToClusterError
+            if there was an issue adding the instance to the cluster.
+
+        Args:
+            instance_address: address of the instance to add to the cluster
+        """
+        options = {
+            "password": self.cluster_admin_password,
+            "recoveryMethod": "clone",
```

Comment:
Should it always `clone`? In the case of a re-adding an instance, can this be potentially very taxing? What's the thoughts against leaving it in `auto` ?

Reply:
I tried to set to `auto`, but got the following error trace:

```
The target instance \'juju-840aa8-2:3306\' has not been pre-provisioned (GTID set is empty). The Shell is unable to decide whether incremental state recovery can correctly provision it.
The safest and most convenient way to provision a new instance is through automatic clone provisioning, which will completely overwrite the state of \'juju-840aa8-2:3306\' with a physical snapshot from an existing cluster member. To use this method by default, set the \'recoveryMethod\' option to \'clone\'.

The incremental state recovery may be safely used if you are sure all updates ever executed in the cluster were done with GTIDs enabled, there are no purged transactions and the new instance contains the same GTID set as the cluster or a subset of it. To use this method by default, set the \'recoveryMethod\' option to \'incremental\'.

Traceback (most recent call last):
  File "<string>", line 3, in <module>
RuntimeError: Cluster.add_instance: \'recoveryMethod\' option must be set to \'clone\' or \'incremental\'
```

Reply:
That's interesting. @delgod illuminate us with your thoughts

Reply:
I think we should try `auto` first. and switch to `clone` only if `auto` doesn't work

Reply:
addressed in [aec0bd5](https://github.com/canonical/mysql-operator/pull/6/commits/aec0bd5f1ba0151b639a94eb6959bbc4ed6c7349). we try `auto` first, and if there is an exception, we try `clone`